### PR TITLE
swap keys and values in preload dict

### PR DESCRIPTION
**Pull request by snopoke**
http://manage.dimagi.com/default.asp?162213

@TylerSheffels

### CODE REVIEW

File: corehq/apps/app_manager/models.py [hunk 1]
```diff
@@ -367,7 +367,7 @@ class AutoSelectCase(DocumentSchema):
 class LoadUpdateAction(AdvancedAction):
     """
     details_module:     Use the case list configuration from this module to show the cases.
-    preload:            Value from the case to load into the form.
+    preload:            Value from the case to load into the form. Keys are question paths, values are case properties.
```
Comment (dimagimon) [inline corehq/apps/app_manager/models.py]: E501 line too long (119 > 115 characters)

File: corehq/apps/app_manager/models.py [hunk 2]
```diff
@@ -1855,6 +1855,18 @@ class AdvancedForm(IndexedFormBase, NavMenuItemMediaMixin):
...
 
+    @classmethod
+    def wrap(cls, data):
+        # lazy migration to swap keys with values in action preload dict.
+        # http://manage.dimagi.com/default.asp?162213
+        load_actions = data.get('actions', {}).get('load_update_cases', [])
+        for action in load_actions:
+            preload = action['preload']
+            if preload and preload.values()[0].startswith('/'):
+                action['preload'] = {v: k for k, v in preload.items()}
```
Comment (snopoke) [inline corehq/apps/app_manager/models.py]: this is safe since there is already validation in the UI that prevents duplicate values

**Comment by snopoke** [comment]
@dannyroberts

**Comment by dannyroberts** [comment]
@snopoke was the original issue that advanced module had the key and value backwards from normal module when it came to preloads?

**Comment by snopoke** [comment]
Yes

File: corehq/apps/app_manager/management/commands/migrate_advanced_form_preload.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
+from corehq.apps.app_manager.management.commands.migrate_graph_configuration_properties import AppMigrationCommandBase
```
Comment (dannyroberts) [inline corehq/apps/app_manager/management/commands/migrate_advanced_form_preload.py]: This is great, pulling out the boilerplate. Would be nice to put `AppMigrationCommandBase` somewhere other than `migrate_graph_configuration_properties`, like `corehq.apps.app_manager.management.helpers`, but I certainly won't block the PR on that

Comment (snopoke) [inline corehq/apps/app_manager/management/commands/migrate_advanced_form_preload.py]: :+1:

File: corehq/apps/app_manager/static/app_manager/js/case-config-utils.js [hunk 4]
```diff
@@ -47,5 +47,34 @@ var CC_UTILS = {
...
+            var path = case_property[!keyIsPath ? 'path' : 'key'];
+            if (key || path) {
+                if (_(required).contains(key) && case_property.required) {
+                    extra_dict[key] = path;
+                } else {
+                    property_dict[key] = path;
+                }
+            }
+        });
+        return [property_dict, extra_dict];
```
Comment (dannyroberts) [inline corehq/apps/app_manager/static/app_manager/js/case-config-utils.js]: Nice

**Comment by dannyroberts** [comment]
Awesome, looks good to me.

File: corehq/apps/app_manager/xform.py [hunk 5]
```diff
@@ -1421,7 +1421,7 @@ def check_case_type(action):
             session_case_id = CaseIDXPath(session_var(var_name))
             if action.preload:
                 self.add_casedb()
-                for property, nodeset in action.preload.items():
+                for nodeset, property, in action.preload.items():
```
Comment (dannyroberts) [inline corehq/apps/app_manager/xform.py]: Just noticed funny syntax with trailing `,`

**Pull request merged**